### PR TITLE
create meta package [RFC]

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

Creating meta-package Lightning as an umbrella for `pytorch` and `app`

this support installing from source as well as building packages for:
- `lightning_app` as standalone
- `pytorch_lightning` as standalone
- `lightning` including both above + meta package

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/Lightning-AI/lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @tchaton @rohitgr7 @justusschock @awaelchli @borda

### CODE REVIEW

**Comment by Borda** [comment]
This is the minimal structure with several Todo which shall be addressed in following PRs
the major missing please is updating the Release workflow, but lets keep this PR focused :rabbit:

File: src/lightning_app/__init__.py [hunk 1]
```diff
@@ -19,8 +19,13 @@
     _logger.propagate = False
 
 
+from lightning_app import __about__  # noqa: E402
 from lightning_app import components  # noqa: E402, F401
 from lightning_app.__about__ import *  # noqa: E402, F401, F403
+
+if not hasattr(__about__, "__version__"):
```
Comment (tchaton) [inline src/lightning_app/__init__.py:26]: Is it to resolve a bug with the meta package generator ?

File: src/pytorch_lightning/strategies/bagua.py [hunk 2]
```diff
@@ -20,10 +20,12 @@
 from pytorch_lightning.trainer.states import TrainerFn
 from pytorch_lightning.utilities.distributed import ReduceOp
 from pytorch_lightning.utilities.exceptions import MisconfigurationException
-from pytorch_lightning.utilities.imports import _BAGUA_AVAILABLE
+from pytorch_lightning.utilities.imports import _package_available
 from pytorch_lightning.utilities.optimizer import optimizers_to_device
 from pytorch_lightning.utilities.seed import reset_seed
 
+_BAGUA_AVAILABLE = _package_available("bagua")
```
Comment (tchaton) [inline src/pytorch_lightning/strategies/bagua.py:27]: Why it is required ?

File: src/pytorch_lightning/strategies/sharded_spawn.py [hunk 3]
```diff
@@ -19,10 +19,10 @@
 from torch.optim import Optimizer
 
 import pytorch_lightning as pl
+from pytorch_lightning.overrides.fairscale import _FAIRSCALE_AVAILABLE
```
Comment (tchaton) [inline src/pytorch_lightning/strategies/sharded_spawn.py:22]: Did you make the change to get the full import path?

Comment (tchaton) [reply src/pytorch_lightning/strategies/sharded_spawn.py:22]: If yes, it means we need to run the meta-package on each commit to verify the imports won't break for a release.

**Review by tchaton — APPROVED** [review]
Overall LGTM. I will have to play with it though to get a better feeling of its un-stablity.

File: .github/workflows/ci_pkg-install.yml [hunk 4]
```diff
@@ -32,38 +40,49 @@ jobs:
...
-          # this is just a hotfix because of Win cannot install it directly
-          pip install -r requirements.txt --find-links https://download.pytorch.org/whl/cpu/torch_stable.html
+      # ToDo
+      #- name: Setup Windows
+      #  if: runner.os == 'windows'
+      #  run: |
+      #    # this is just a hotfix because of Win cannot install it directly
+      #    pip install -r requirements.txt --find-links https://download.pytorch.org/whl/cpu/torch_stable.html
+
+      - name: Package name
```
[the remark this answers is not in the document]
Comment (Borda) [reply .github/workflows/ci_pkg-install.yml:68]: The name inside lightning and as stand alone package is different

File: .github/workflows/ci_pkg-install.yml [hunk 5]
```diff
@@ -32,38 +40,49 @@ jobs:
...
 
       - name: Install | Uninstall package - wheel
         working-directory: ./dist
         run: |
           # install as wheel
           pip install *.whl
-          cd ..
-          python -c "import pytorch_lightning ; print(pytorch_lightning.__version__)"
-          pip uninstall -y pytorch_lightning
+          pip list | grep lightning
```
[the remark this answers is not in the document]
Comment (Borda) [reply .github/workflows/ci_pkg-install.yml:86]: Nop =)

File: src/pytorch_lightning/strategies/bagua.py (hunk 2 above)
Comment (Borda) [reply src/pytorch_lightning/strategies/bagua.py:27]: Because the class is strangely defined in if statement and actual meta pkg creation canot resolve it, BUT it is also the same way we already set for Deepspeed so it is consistent

File: src/pytorch_lightning/strategies/sharded_spawn.py (hunk 3 above)
Comment (Borda) [reply src/pytorch_lightning/strategies/sharded_spawn.py:22]: Yes, it runs as part of the CI

File: src/lightning_app/__init__.py (hunk 1 above)
Comment (Borda) [reply src/lightning_app/__init__.py:26]: Not bug, but compatibility as in future we move version to separate file

File: src/pytorch_lightning/__about__.py [hunk 6]
```diff
@@ -1,12 +1,25 @@
...
-__author__ = "William Falcon et al."
-__author_email__ = "waf2107@columbia.edu"
+# __version__ = "1.7.0dev"
+__author__ = "Lightning AI et al."
+__author_email__ = "pytorch@lightning.ai"
 __license__ = "Apache-2.0"
-__copyright__ = f"Copyright (c) 2018-{_this_year}, {__author__}."
-__homepage__ = "https://github.com/PyTorchLightning/pytorch-lightning"
+__copyright__ = f"Copyright (c) 2018-{time.strftime('%Y')}, {__author__}."
+__homepage__ = "https://github.com/Lightning-AI/pytorch-lightning"
```
Comment (otaj) [inline src/pytorch_lightning/__about__.py:22]:
```suggestion
__homepage__ = "https://github.com/Lightning-AI/lightning"
```

**Review by otaj — APPROVED** [review]
LGTM, left minor comments

**Review by ethanwharris — APPROVED** [review]

File: .github/workflows/ci_pkg-install.yml [hunk 7]
```diff
@@ -22,7 +26,11 @@ jobs:
       # max-parallel: 6
       matrix:
         os: [ubuntu-20.04, macOS-10.15, windows-2019]
-        python-version: [3.7, 3.9]
+        pkg: ["", "app", "pytorch"]  # todo: add following use-case "lightning" after resowing pulling sources
+        # in fact, lest make this se depending jobs, at fist build app and pytorch, and save these packages as artefact
+        # then, in the following job pull artefacts and install these packages and build and test lightning
+        # todo: in such case abstract this pkg testing as actions to reduce variation in testing
```
Comment (manskx) [inline .github/workflows/ci_pkg-install.yml:32]: minor typos:

Comment (Borda) [reply .github/workflows/ci_pkg-install.yml:32]: which?

File: src/lightning/__about__.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+# limitations under the License.
+
+import time
+
+__author__ = "Lightning AI et al."
+__author_email__ = "pytorch@lightning.ai"
+__license__ = "Apache-2.0"
+__copyright__ = f"Copyright (c) 2018-{time.strftime('%Y')}, {__author__}."
+__homepage__ = "https://github.com/Lightning-AI/pytorch-lightning"
+__docs_url__ = "https://pytorch-lightning.readthedocs.io/en/stable/"
```
Comment (manskx) [inline src/lightning/__about__.py:23]: # TODO: combined docs

File: src/lightning/__init__.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+from lightning.pytorch.core import LightningDataModule, LightningModule  # noqa: E402
+from lightning.pytorch.trainer import Trainer  # noqa: E402
+from lightning.pytorch.utilities.seed import seed_everything  # noqa: E402
+
+__all__ = [
+    "LightningApp",
+    "LightningFlow",
+    "LightningWork",
+    "BuildConfig",
+    "CloudCompute",
```
Comment (manskx) [inline src/lightning/__init__.py:51]: do we want to add the `Drive` here?

Comment (Borda) [reply src/lightning/__init__.py:51]: this is state of DevCon, it does not represent the latest state which needs to be posted later this week :rabbit:

File: src/lightning/__about__.py (hunk 8 above)
Comment (Borda) [reply src/lightning/__about__.py:23]: good one :)

File: MANIFEST.in [hunk 10]
```diff
@@ -1,5 +1,5 @@
-include pytorch_lightning/py.typed  # marker file for PEP 561
```
Comment (carmocca) [inline MANIFEST.in:1]: This was kept on purpose. Why remove it?

File: MANIFEST.in [hunk 11]
```diff
@@ -1,5 +1,5 @@
...
-include CHANGELOG.md
```
Comment (carmocca) [inline MANIFEST.in:2]: It's a good idea to distribute the CHANGELOG. Why remove it?

File: MANIFEST.in [hunk 12]
```diff
@@ -1,5 +1,5 @@
...
-recursive-include requirements *.txt
-include requirements.txt
+exclude *.toml  # project config
+exclude requirements.txt
+exclude __pycache__
```
Comment (carmocca) [inline MANIFEST.in:3]: This shouldn't be necessary

File: MANIFEST.in [hunk 13]
```diff
@@ -1,5 +1,5 @@
-include pytorch_lightning/py.typed  # marker file for PEP 561
-include CHANGELOG.md
-recursive-include requirements *.txt
-include requirements.txt
+exclude *.toml  # project config
+exclude requirements.txt
```
Comment (carmocca) [inline MANIFEST.in:2]: It's a good idea to distribute the requirements. Why remove it? Also it needs to be done recursively.

File: MANIFEST.in [hunk 14]
```diff
@@ -1,5 +1,5 @@
...
+exclude __pycache__
+include .actions/setup_tools.py
```
Comment (carmocca) [inline MANIFEST.in:4]: Why do we need to include this script in the release when it's only used by us at build time?

File: requirements/pytorch/base.txt [hunk 15]
```diff
@@ -4,7 +4,7 @@ tqdm>=4.57.0, <=4.63.0
 PyYAML>=5.4, <=6.0
 fsspec[http]>=2021.05.0, !=2021.06.0, <2022.6.0
 tensorboard>=2.2.0, <2.10.0
-torchmetrics>=0.4.1, <0.9.2
+torchmetrics>=0.7.0, <0.9.2  # needed for using fixed compare_version
```
Comment (carmocca) [inline requirements/pytorch/base.txt:7]: This is an important breaking change and should be mentioned in the CHANGELOG

File: src/lightning/__init__.py [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+
+from lightning.__about__ import *  # noqa: E402, F401, F403
+from lightning.__version__ import version as __version__  # noqa: E402, F401
+from lightning.app.core.app import LightningApp  # noqa: E402
+from lightning.app.core.flow import LightningFlow  # noqa: E402
+from lightning.app.core.work import LightningWork  # noqa: E402
+from lightning.app.utilities.packaging.build_config import BuildConfig  # noqa: E402
+from lightning.app.utilities.packaging.cloud_compute import CloudCompute  # noqa: E402
+from lightning.pytorch.callbacks import Callback  # noqa: E402
+from lightning.pytorch.core import LightningDataModule, LightningModule  # noqa: E402
```
Comment (carmocca) [inline src/lightning/__init__.py:42]: Shouldn't these be gated by `if _PL_AVAILABLE` and same for apps?

File: src/lightning/__init__.py [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+    if self.isEnabledFor(_DETAIL):
+        # logger takes its '*args' as 'args'
+        self._log(_DETAIL, message, args, **kwargs)
+
+
+logging.addLevelName(_DETAIL, "DETAIL")
+logging.detail = _detail
+logging.Logger.detail = _detail
+
+_root_logger = logging.getLogger()
```
Comment (carmocca) [inline src/lightning/__init__.py:18]: I don't think that the wrapper should be defining logging logic. Where would we use it? Why?

File: src/lightning/__setup__.py [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+from typing import Any, Dict
+
+from setuptools import find_packages
+
+_PROJECT_ROOT = "."
+_SOURCE_ROOT = os.path.join(_PROJECT_ROOT, "src")
+_PACKAGE_ROOT = os.path.join(_SOURCE_ROOT, "lightning")
+
+
+def _load_py_module(name: str, location: str) -> ModuleType:
```
Comment (carmocca) [inline src/lightning/__setup__.py:15]: Isn't this already in setup_tools?

**Review by carmocca — COMMENTED** [review]
Not very happy with this PR. There are many changes that should have been done separately and I don't understand some of the logic and comments leftover.

I still wonder if we really need such a complex and hacky solution for the wrapper.

File: MANIFEST.in (hunk 11 above)
Comment (Borda) [reply MANIFEST.in:2]: it is there: https://github.com/Lightning-AI/lightning/blob/55e347ba609df1a1248ef1243ff8cbc09fb97bc8/src/pytorch_lightning/__setup__.py#L56

File: MANIFEST.in (hunk 10 above)
Comment (Borda) [reply MANIFEST.in:1]: it is there: https://github.com/Lightning-AI/lightning/blob/55e347ba609df1a1248ef1243ff8cbc09fb97bc8/src/pytorch_lightning/__setup__.py#L54

File: MANIFEST.in (hunk 12 above)
Comment (Borda) [reply MANIFEST.in:3]: seems it did not do anything in the end :innocent:

File: MANIFEST.in (hunk 14 above)
Comment (Borda) [reply MANIFEST.in:4]: it turned out that when you install a package from a build package it is also called setup and then it is needed, but we can bypass it by adding an extra file in the package indication we run the prepared package...

File: src/lightning/__init__.py (hunk 16 above)
Comment (Borda) [reply src/lightning/__init__.py:42]: not sure why? :rabbit:

File: src/lightning/__init__.py (hunk 17 above)
Comment (Borda) [reply src/lightning/__init__.py:18]: I am fine to prune it, just did the union of both init...

File: src/lightning/__setup__.py (hunk 18 above)
Comment (Borda) [reply src/lightning/__setup__.py:15]: yes but you need it to be able to load `setup_tools` :)

**Comment by Borda** [comment]
> I still wonder if we really need such a complex and hacky solution for the wrapper.

in other words, you are asking why we have two packages in one repo?
